### PR TITLE
Improve search api

### PR DESCRIPTION
**Pull request by sk274**
## Why was this change made?
Improve search api
- exact phrase search with 'simple' language configuration (#175)
- normalize diacritics for searches (#177)
- rank search results based on index weights (main title > subtitle > uri > rest) (#178)
- treat lone * query same as no argument to search, returning all (#196)
- search with Blue Core URI (#198)

## How was this change tested?
Locally via unit test & bluecore-stack

## Which documentation and/or configurations were updated?
N/A

### CODE REVIEW

File: src/bluecore_api/app/routes/search.py [hunk 1]
```diff
@@ -21,36 +20,66 @@
...
 def format_query(query: str) -> str:
     """
     Format query
-    Replace multiple spaces with a single space.
-    Separate any phrase terms inside double quotes with <-> and strip double quotes.
-    Leave ' | ' sequences intact for OR operations.
-    Replace any other spaces with ' & ' for AND operations.
-    Replace '*' with ':*' for wildcard search.
+        - Remove leading and trailing spaces.
+        - Remove invalid wildcards (e.g. "hello * world *" -> "hello world ").
```
Comment (JeremyDuncan) [inline src/bluecore_api/app/routes/search.py:33]: Would this be an issue if a wildcard was part of a books title? I'm just thinking of edge cases. for example:
* `M*A*S*H` series - https://en.wikipedia.org/wiki/MASH:_A_Novel_About_Three_Army_Doctors
* "***" (Thats the title) -  https://en.wikipedia.org/wiki/***_(novel)

Comment (sk274) [reply src/bluecore_api/app/routes/search.py:33]: I couldn't find a good way to distinguish whether a user wanted to use a wildcard or asterisk as string literal.
If you have any good ideas, please let me know!

Comment (JeremyDuncan) [reply src/bluecore_api/app/routes/search.py:33]: Hmm this is a conundrum lol.

Some ideas (though may add complexity) would possibly adding some fallback logic that uses a LIKE query with literal string containing any "\*" IF no results are found with initial query logic and there was a "\*" in the query.

Example:
```
if not results and "*" in original_query:
    stmt = stmt.where(ResourceBase.data.cast(String).ilike(f"%{original_query}%"))
```


---

Another idea would be:
Allowing users to escape "\*" with a backslash (example: `M\*A\*S\*H`) and treat that query literally
This would require keeping the user aware of the "*" nuance in a query note somewhere and explaining how to get around it with "\\".

**Comment by JeremyDuncan** [comment]
Other than my one edge case scenario I mentioned, everything else looks good to me, and it passes with all the integration tests too!

File: src/bluecore_api/app/routes/search.py (hunk 1 above)
Comment (sk274) [reply src/bluecore_api/app/routes/search.py:33]: Thanks for your review!

I am afraid the fallback could be very expensive - we don't have index on the data column and I am not sure if we want to create one.

I was on transit and didn't give you the full context, I am sorry about that!
The more general problem is with the way the special characters are handled in the tsvector.
I could be wrong but I think to_tsvector (for indexing) and to_tsquery (for querying) treat most special characters as word separators and are removed, even with the 'simple' configuration.
I tested out simple sql queries like this and the output appears to back that up.
```
select to_tsvector('english', 'MAS*H'); -> 'h':2 'mas':1
select to_tsvector('english', 'MAS\*H'); -> 'h':2 'mas':1
select to_tsvector('english', 'MAS:H'); -> 'h':2 'mas':1
select to_tsvector('english', 'MAS|H'); ->  'h':2 'mas':1
---------------
select to_tsquery('english', 'MAS*H'); -> 'mas' <-> 'h'
select to_tsquery('english', 'MAS\*H'); -> 'mas' <-> 'h'
select to_tsquery('english', 'MAS&H'); -> 'mas' & 'h'
select to_tsquery('english', 'MAS\&H'); -> 'mas' <-> 'h'
select to_tsquery('english', 'D♭ & major.'); -> 'd' & 'major'
```

To properly support indexing/searching special characters, I think we will need to transform them for indexing, e.g. MAS*H -> MAS__ASTERISK__H and treat user query accordingly.
#177 states that the symbols maps aren't part of the MVP for any participating libraries so I didn't try to implement that in this sprint.

**Review by JeremyDuncan — APPROVED** [review]
🙌

**Review by jermnelson — APPROVED** [review]
